File: labs/lab09_carry_lookahead/carry_lookahead_adder.py
#!/usr/bin/env python3
"""Lab 9 – Carry Look-Ahead Adder (CLA) simulation.

Computes all carries in parallel using Generate and Propagate signals,
then produces all sum bits simultaneously—eliminating the ripple delay.

Run:
  python3 carry_lookahead_adder.py              # demo
  python3 carry_lookahead_adder.py <A> <B> [bits]
  Example: python3 carry_lookahead_adder.py 13 11 4
"""

import sys


# ---------------------------------------------------------------------------
# CLA core
# ---------------------------------------------------------------------------

def carry_lookahead_adder(a: int, b: int, n: int, cin: int = 0):
    """Add n-bit numbers using carry look-ahead.

    Returns (sum_int, carries, generate_bits, propagate_bits, overflow).
    """
    # Step 1 – compute per-bit Generate (G) and Propagate (P)
    G = [(( a >> i) & 1) & ((b >> i) & 1) for i in range(n)]   # G_i = A_i AND B_i
    P = [(( a >> i) & 1) ^ ((b >> i) & 1) for i in range(n)]   # P_i = A_i XOR B_i

    # Step 2 – look-ahead carry computation
    # C_0 = cin
    # C_{i+1} = G_i OR (P_i AND C_i)
    C = [0] * (n + 1)
    C[0] = cin
    for i in range(n):
        C[i + 1] = G[i] | (P[i] & C[i])

    # Step 3 – sum bits
    # S_i = P_i XOR C_i  (same as A_i XOR B_i XOR C_i)
    S = [P[i] ^ C[i] for i in range(n)]

    result   = sum(S[i] << i for i in range(n))
    overflow = C[n]

    return result, C, G, P, overflow


# ---------------------------------------------------------------------------
# Display helpers
# ---------------------------------------------------------------------------

def int_to_bits(val: int, n: int) -> list[int]:
    return [(val >> i) & 1 for i in range(n)]


def fmt(bits: list[int]) -> str:
    return "".join(str(b) for b in reversed(bits))


def print_cla(a: int, b: int, n: int, cin: int = 0):
    result, C, G, P, overflow = carry_lookahead_adder(a, b, n, cin)

    a_bits = int_to_bits(a, n)
    b_bits = int_to_bits(b, n)
    s_bits = int_to_bits(result, n)

    print(f"\n{n}-bit Carry Look-Ahead Addition")
    print(f"  A = {fmt(a_bits)}  ({a})")
    print(f"  B = {fmt(b_bits)}  ({b})")
    if cin:
        print(f"  Cin = {cin}")
    print()

    print(f"  {'Bit':>4}  {'A':>2}  {'B':>2}  {'G':>2}  {'P':>2}  {'Cin':>4}  {'Sum':>4}  {'Cout':>5}")
    print(f"  {'─'*38}")
    for i in range(n):
        ai = (a >> i) & 1
        bi = (b >> i) & 1
        print(f"  {i:>4}  {ai:>2}  {bi:>2}  {G[i]:>2}  {P[i]:>2}  {C[i]:>4}  {s_bits[i]:>4}  {C[i+1]:>5}")

    print(f"\n  Sum  = {fmt(s_bits)}  ({result})")
    print(f"  Cout = {overflow}  {'← overflow!' if overflow else ''}")

    # Show the look-ahead carry equations
    print("\n  Look-Ahead Carry Equations (parallel, no ripple):")
    for i in range(1, n + 1):
        terms = [f"G{i-1}"]
        prop  = f"P{i-1}"
        for j in range(i - 1, 0, -1):
            prop_chain = "".join(f"P{k}" for k in range(j, i))
            terms.append(f"{prop_chain}·G{j-1}")
        prop_all = "".join(f"P{k}" for k in range(i))
        terms.append(f"{prop_all}·C0")
        print(f"  C{i} = " + " + ".join(terms))


def demo():
    print("=" * 50)
    print("Carry Look-Ahead Adder Demo")
    print("=" * 50)
    print_cla(5, 3, 4)
    print_cla(13, 11, 4)
    print_cla(170, 85, 8)


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main():
    if len(sys.argv) >= 3:
        a = int(sys.argv[1])
        b = int(sys.argv[2])
        n = int(sys.argv[3]) if len(sys.argv) >= 4 else max(a, b).bit_length() + 1
        print_cla(a, b, n)
    else:
        demo()


if __name__ == "__main__":
    main()

File: labs/lab09_carry_lookahead/carry_lookahead_adder.cpp
// Lab 09 – Carry Look-Ahead Adder (CLA)
// Computes all carries in parallel using Generate and Propagate signals.
// Input: A B n
// Output: G/P vectors, look-ahead carry equations, sum

#include <iostream>
#include <vector>
#include <string>
using namespace std;

struct CLAResult {
    int sum;
    int carryOut;
    vector<int> G, P, C;   // C has n+1 elements: C[0]=cin, C[n]=carry-out
};

CLAResult cla(int A, int B, int n, int cin = 0) {
    vector<int> G(n), P(n), C(n+1);
    C[0] = cin;

    for (int i = 0; i < n; i++) {
        int ai = (A >> i) & 1;
        int bi = (B >> i) & 1;
        G[i] = ai & bi;           // Generate
        P[i] = ai ^ bi;           // Propagate
    }
    for (int i = 0; i < n; i++)
        C[i+1] = G[i] | (P[i] & C[i]);

    int sum = 0;
    for (int i = 0; i < n; i++)
        sum |= (P[i] ^ C[i]) << i;

    return { sum, C[n], G, P, C };
}

string toBinary(int val, int n) {
    string s(n, '0');
    for (int i = 0; i < n; i++)
        s[n-1-i] = '0' + ((val >> i) & 1);
    return s;
}

int main() {
    int A, B, n;
    cout << "Enter A B n (e.g. 13 11 4): ";
    cin >> A >> B >> n;

    auto res = cla(A, B, n);

    cout << "\n" << n << "-bit Carry Look-Ahead Addition\n";
    cout << "  A = " << toBinary(A, n) << "  (" << A << ")\n";
    cout << "  B = " << toBinary(B, n) << "  (" << B << ")\n\n";

    cout << "  Bit   A  B   G  P  Cin  Sum  Cout\n";
    cout << "  " << string(38, '-') << "\n";
    for (int i = 0; i < n; i++) {
        int ai = (A >> i) & 1, bi = (B >> i) & 1;
        int si = res.P[i] ^ res.C[i];
        cout << "   " << i << "    " << ai << "  " << bi
             << "   " << res.G[i] << "  " << res.P[i]
             << "    " << res.C[i] << "    " << si
             << "     " << res.C[i+1] << "\n";
    }

    cout << "\n  Sum  = " << toBinary(res.sum, n) << "  (" << res.sum << ")\n";
    cout << "  Cout = " << res.carryOut;
    if (res.carryOut) cout << "  <- overflow!";
    cout << "\n";

    cout << "\nLook-Ahead Carry Equations (no ripple):\n";
    for (int i = 1; i <= n; i++) {
        cout << "  C" << i << " = G" << (i-1);
        string prop = "";
        for (int j = i-1; j >= 1; j--) {
            prop += "P" + to_string(j-1);
            // build product of P[j-1..i-2] then G[j-1]
            // simplified: just note the structure
        }
        // Print full expansion
        for (int j = i-2; j >= 0; j--) {
            cout << " + (";
            for (int k = j+1; k < i; k++) cout << "P" << k-1;
            cout << "*G" << j << ")";
        }
        cout << " + (";
        for (int k = 0; k < i; k++) cout << "P" << k;
        cout << "*C0)\n";
    }
    return 0;
}
